extensions/amp-3q-player/0.1/amp-3q-player.js

```diff
@@ -0,0 +1,190 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {tryParseJson} from '../../../src/json';
+import {user} from '../../../src/log';
+import {removeElement} from '../../../src/dom';
+import {
+  installVideoManagerForDoc,
+} from '../../../src/service/video-manager-impl';
+import {isObject} from '../../../src/types';
+import {VideoEvents} from '../../../src/video-interface';
+import {videoManagerForDoc} from '../../../src/video-manager';
+
+/**
+ * @implements {../../../src/video-interface.VideoInterface}
+ */
+class Amp3QPlayer extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Element} */
+    this.iframe_ = null;
+
+    /** @private {?Promise} */
+    this.playerReadyPromise_ = null;
```

Comment:
Unless your video player does not accept postMessages until it send an initial postMessage back to the component (like amp-youtube) there is no need for a separate `playerReadyResolver_`. This can become:

```
/** @private {?Promise} */
this.playerReadyPromise_ = null;
```

in `layoutCallback`:
```
this.playerReadyPromise_ = this.loadPromise(this.iframe_).....
return this.playerReadyPromise_;
```